### PR TITLE
Add check and cleanup operation on project permisions

### PR DESCRIPTION
**Pull request by autumnfound**
Only removes users that are directly set on projects that are not bots
or not admins.

Signed-off-by: Martin Lowe <martin.lowe@eclipse-foundation.org>

### CODE REVIEW

**Review by chrisguindon — APPROVED** [review]
LGTM

**Comment by autumnfound** [comment]
@fredg02 any concerns merging this in today?

**Comment by fredg02** [comment]
No concerns. Can you test this with a single organization yet?

**Comment by autumnfound** [comment]
> No concerns. Can you test this with a single organization yet?

Not yet, but it is something that I am planning on adding. You've reminded me that I should add that as an issue. If you like, we can postpone this merge while I do that? It's not an overly difficult change to do so it should be done in the next day or two at most.

**Comment by autumnfound** [comment]
I've created an issue to track adding new params: https://github.com/EclipseFdn/eclipsefdn-github-sync/issues/150

**Comment by fredg02** [comment]
:+1: for postponing the merge and testing it on a single org.

**Comment by autumnfound** [comment]
@fredg02 I've merged in the change to allow for single project targeted runs of the script. You need the short project ID of the project you wish to target, and pass it via either the p or P param (depends on script, you can check top of file to check or use --help). I've rebased this branch to include that change, so you can check it out and play with it if you like

**Comment by fredg02** [comment]
Thanks. My original question if this can be tested with a single organization was intended to be a recommendation to have a general policy of a staged rollout. Potential issues/regressions are then limited to a single org instead of "all the orgs".

**Comment by fredg02** [comment]
So if the changes in this PR work as expected with a single org, then +1 for merging it.

**Comment by autumnfound** [comment]
I've tested this locally with the following command and getting the output below. Let me know if there are any concerns, otherwise I'll merge this in and push it to production next week. Thanks!

node src/GitlabSync.js -P dash -H <redacted> -V -s <redacted>
TLDR: `2021-04-23T15:24:35 [main] INFO Removing user 'webmaster' from project 'org.eclipse.dash.handbook'(1681)`

```shell
2021-04-23T15:24:34 [main] DEBUG GitlabSync:sanitizeGroupName(pid = eclipse)
2021-04-23T15:24:34 [main] DEBUG GitlabSync:sanitizeGroupName(pid = dash)
2021-04-23T15:24:34 [main] DEBUG GitlabSync:sanitizeGroupName(pid = import-test)
2021-04-23T15:24:34 [main] DEBUG GitlabSync:sanitizeGroupName(pid = t-migration-1610-group)
2021-04-23T15:24:34 [main] DEBUG GitlabSync:sanitizeGroupName(pid = test)
2021-04-23T15:24:34 [main] DEBUG GitlabSync:sanitizeGroupName(pid = tmp-migration-1612-group)
2021-04-23T15:24:34 [main] INFO Starting sync
2021-04-23T15:24:34 [main] DEBUG GitlabSync:getGroup(name = Eclipse, path = eclipse, parent = undefined, visibility = public)
2021-04-23T15:24:34 [main] DEBUG GitlabSync:sanitizeGroupName(pid = eclipse)
2021-04-23T15:24:34 [main] INFO Project target set ('dash'). Skipping non-matching project ID app4mc
2021-04-23T15:24:34 [main] INFO Project target set ('dash'). Skipping non-matching project ID simopenpass
2021-04-23T15:24:34 [main] INFO Project target set ('dash'). Skipping non-matching project ID hawk
2021-04-23T15:24:34 [main] INFO Project target set ('dash'). Skipping non-matching project ID chess
2021-04-23T15:24:34 [main] INFO Project target set ('dash'). Skipping non-matching project ID ice
2021-04-23T15:24:34 [main] INFO Project target set ('dash'). Skipping non-matching project ID asciidoc
2021-04-23T15:24:34 [main] INFO Project target set ('dash'). Skipping non-matching project ID austen
2021-04-23T15:24:34 [main] INFO Project target set ('dash'). Skipping non-matching project ID comma
2021-04-23T15:24:34 [main] INFO Processing 'dash'
2021-04-23T15:24:34 [main] DEBUG GitlabSync:getGroup(name = Eclipse Dash, path = dash, parent = [object Object], visibility = public)
2021-04-23T15:24:34 [main] DEBUG GitlabSync:sanitizeGroupName(pid = dash)
2021-04-23T15:24:34 [main] DEBUG GitlabSync:getUserList(project = {"project_id":"technology.dash","short_project_id":"dash","name":"Eclipse Dash","summary":"Eclipse Dash is a place where the community itself collaborates on tools for community awareness and collaboration in support of our ultimate objective of committer quality and cooperation....","url":"https://projects.eclipse.org/projects/technology.dash","website_url":"http://www.eclipse.org/dash/","logo":"https://projects.eclipse.org/sites/all/modules/custom/pmi/project_state/images/incubating.png","tags":[],"github_repos":[{"url":"https://github.com/eclipse/dash-licenses"},{"url":"https://github.com/eclipse/dash-website"}],"gitlab_repos":[{"url":"https://gitlab.eclipse.org/eclipse/dash/dash-gitlab-testing","org":"dash","repo":"dash-gitlab-testing"},{"url":"https://gitlab.eclipse.org/eclipse/dash/org.eclipse.dash.handbook","org":"dash","repo":"org.eclipse.dash.handbook"}],"gerrit_repos":[{"url":"/gitroot/dash/org.eclipse.dash.dashboard.git"},{"url":"/gitroot/dash/org.eclipse.dash.handbook.git"},{"url":"/gitroot/dash/org.eclipse.dash.m4e.tools.git"},{"url":"/gitroot/dash/org.eclipse.dash.maven.git"}],"contributors":[],"committers":[{"username":"malowe","url":"https://api.eclipse.org/account/profile/malowe"},{"username":"mward","url":"https://api.eclipse.org/account/profile/mward"},{"username":"mdelgado624","url":"https://api.eclipse.org/account/profile/mdelgado624"},{"username":"shmacdonald","url":"https://api.eclipse.org/account/profile/shmacdonald"},{"username":"cguindon","url":"https://api.eclipse.org/account/profile/cguindon"},{"username":"mbarbero","url":"https://api.eclipse.org/account/profile/mbarbero"},{"username":"wbeaton","url":"https://api.eclipse.org/account/profile/wbeaton"},{"username":"scorbett","url":"https://api.eclipse.org/account/profile/scorbett"},{"username":"droy","url":"https://api.eclipse.org/account/profile/droy"}],"project_leads":[{"username":"wbeaton","url":"https://api.eclipse.org/account/profile/wbeaton"}],"working_groups":[],"spec_project_working_group":[],"state":"Incubating","releases":[{"name":"4.0 (Handbook)","url":"https://projects.eclipse.org/projects/technology.dash/releases/4.0-handbook"},{"name":"3.0.0","url":"https://projects.eclipse.org/projects/release/3.0.0-11"}],"pp_repos":["dash-gitlab-testing","org.eclipse.dash.handbook"],"pp_orgs":["dash"]})
2021-04-23T15:24:34 [main] DEBUG GitlabSync:getUser(uname = malowe, url = https://api.eclipse.org/account/profile/malowe)
2021-04-23T15:24:34 [main] DEBUG GitlabSync:addUserToGroup(user = [object Object], group = [object Object], perms = 30)
2021-04-23T15:24:34 [main] DEBUG GitlabSync:getGroupMembers(group = [object Object])
2021-04-23T15:24:35 [main] VERBOSE User 'malowe' is already a member of Eclipse Dash
2021-04-23T15:24:35 [main] INFO Fixing permission level for user 'malowe' in group 'Eclipse Dash'
2021-04-23T15:24:35 [main] ERROR Bad Request
2021-04-23T15:24:35 [main] WARN Error while fixing permission level for user 'malowe' in group 'Eclipse Dash'
2021-04-23T15:24:35 [main] DEBUG GitlabSync:getUser(uname = mward, url = https://api.eclipse.org/account/profile/mward)
2021-04-23T15:24:35 [main] DEBUG GitlabSync:addUserToGroup(user = [object Object], group = [object Object], perms = 30)
2021-04-23T15:24:35 [main] DEBUG GitlabSync:getGroupMembers(group = [object Object])
2021-04-23T15:24:35 [main] VERBOSE User 'mward' is already a member of Eclipse Dash
2021-04-23T15:24:35 [main] DEBUG GitlabSync:getUser(uname = mdelgado624, url = https://api.eclipse.org/account/profile/mdelgado624)
2021-04-23T15:24:35 [main] DEBUG GitlabSync:addUserToGroup(user = [object Object], group = [object Object], perms = 30)
2021-04-23T15:24:35 [main] DEBUG GitlabSync:getGroupMembers(group = [object Object])
2021-04-23T15:24:35 [main] VERBOSE User 'mdelgado624' is already a member of Eclipse Dash
2021-04-23T15:24:35 [main] DEBUG GitlabSync:getUser(uname = shmacdonald, url = https://api.eclipse.org/account/profile/shmacdonald)
2021-04-23T15:24:35 [main] DEBUG GitlabSync:addUserToGroup(user = [object Object], group = [object Object], perms = 30)
2021-04-23T15:24:35 [main] DEBUG GitlabSync:getGroupMembers(group = [object Object])
2021-04-23T15:24:35 [main] VERBOSE User 'shmacdonald' is already a member of Eclipse Dash
2021-04-23T15:24:35 [main] DEBUG GitlabSync:getUser(uname = cguindon, url = https://api.eclipse.org/account/profile/cguindon)
2021-04-23T15:24:35 [main] DEBUG GitlabSync:addUserToGroup(user = [object Object], group = [object Object], perms = 30)
2021-04-23T15:24:35 [main] DEBUG GitlabSync:getGroupMembers(group = [object Object])
2021-04-23T15:24:35 [main] VERBOSE User 'cguindon' is already a member of Eclipse Dash
2021-04-23T15:24:35 [main] DEBUG GitlabSync:getUser(uname = mbarbero, url = https://api.eclipse.org/account/profile/mbarbero)
2021-04-23T15:24:35 [main] DEBUG GitlabSync:addUserToGroup(user = [object Object], group = [object Object], perms = 30)
2021-04-23T15:24:35 [main] DEBUG GitlabSync:getGroupMembers(group = [object Object])
2021-04-23T15:24:35 [main] VERBOSE User 'mbarbero' is already a member of Eclipse Dash
2021-04-23T15:24:35 [main] DEBUG GitlabSync:getUser(uname = wbeaton, url = https://api.eclipse.org/account/profile/wbeaton)
2021-04-23T15:24:35 [main] DEBUG GitlabSync:addUserToGroup(user = [object Object], group = [object Object], perms = 40)
2021-04-23T15:24:35 [main] DEBUG GitlabSync:getGroupMembers(group = [object Object])
2021-04-23T15:24:35 [main] VERBOSE User 'wbeaton' is already a member of Eclipse Dash
2021-04-23T15:24:35 [main] DEBUG GitlabSync:getUser(uname = scorbett, url = https://api.eclipse.org/account/profile/scorbett)
2021-04-23T15:24:35 [main] DEBUG GitlabSync:addUserToGroup(user = [object Object], group = [object Object], perms = 30)
2021-04-23T15:24:35 [main] DEBUG GitlabSync:getGroupMembers(group = [object Object])
2021-04-23T15:24:35 [main] VERBOSE User 'scorbett' is already a member of Eclipse Dash
2021-04-23T15:24:35 [main] DEBUG GitlabSync:getUser(uname = droy, url = https://api.eclipse.org/account/profile/droy)
2021-04-23T15:24:35 [main] DEBUG GitlabSync:addUserToGroup(user = [object Object], group = [object Object], perms = 30)
2021-04-23T15:24:35 [main] DEBUG GitlabSync:getGroupMembers(group = [object Object])
2021-04-23T15:24:35 [main] VERBOSE User 'droy' is already a member of Eclipse Dash
2021-04-23T15:24:35 [main] DEBUG GitlabSync:getUser(uname = dash-bot, url = undefined)
2021-04-23T15:24:35 [main] ERROR Cannot fetch user information for user 'dash-bot' with no set URL
2021-04-23T15:24:35 [main] VERBOSE Could not retrieve user for UID 'dash-bot', skipping
2021-04-23T15:24:35 [main] DEBUG GitlabSync:removeAdditionalUsers(expectedUsers = [object Object], group = [object Object], projectID = dash)
2021-04-23T15:24:35 [main] DEBUG GitlabSync:getGroupMembers(group = [object Object])
2021-04-23T15:24:35 [main] VERBOSE Checking user 'malowe' access to group 'Eclipse Dash'
2021-04-23T15:24:35 [main] VERBOSE Checking user 'mward' access to group 'Eclipse Dash'
2021-04-23T15:24:35 [main] VERBOSE Checking user 'mdelgado624' access to group 'Eclipse Dash'
2021-04-23T15:24:35 [main] VERBOSE Checking user 'shmacdonald' access to group 'Eclipse Dash'
2021-04-23T15:24:35 [main] VERBOSE Checking user 'cguindon' access to group 'Eclipse Dash'
2021-04-23T15:24:35 [main] VERBOSE Checking user 'mbarbero' access to group 'Eclipse Dash'
2021-04-23T15:24:35 [main] VERBOSE Checking user 'wbeaton' access to group 'Eclipse Dash'
2021-04-23T15:24:35 [main] VERBOSE Checking user 'scorbett' access to group 'Eclipse Dash'
2021-04-23T15:24:35 [main] VERBOSE Checking user 'droy' access to group 'Eclipse Dash'
2021-04-23T15:24:35 [main] DEBUG Processing repo 'https://gitlab.eclipse.org/eclipse/dash/dash-gitlab-testing'
2021-04-23T15:24:35 [main] DEBUG GitlabSync:getProject(name = dash-gitlab-testing, parent = [object Object])
2021-04-23T15:24:35 [main] DEBUG GitlabSync:cleanUpProjectUsers(project = 1680)
2021-04-23T15:24:35 [main] DEBUG Processing repo 'https://gitlab.eclipse.org/eclipse/dash/org.eclipse.dash.handbook'
2021-04-23T15:24:35 [main] DEBUG GitlabSync:getProject(name = org.eclipse.dash.handbook, parent = [object Object])
2021-04-23T15:24:35 [main] DEBUG GitlabSync:cleanUpProjectUsers(project = 1681)
2021-04-23T15:24:35 [main] INFO Removing user 'webmaster' from project 'org.eclipse.dash.handbook'(1681)
2021-04-23T15:24:35 [main] INFO Project target set ('dash'). Skipping non-matching project ID dataeggs
2021-04-23T15:24:35 [main] INFO Project target set ('dash'). Skipping non-matching project ID escet
2021-04-23T15:24:35 [main] INFO Project target set ('dash'). Skipping non-matching project ID lsat
2021-04-23T15:24:35 [main] INFO Project target set ('dash'). Skipping non-matching project ID mpc
2021-04-23T15:24:35 [main] INFO Project target set ('dash'). Skipping non-matching project ID set
2021-04-23T15:24:35 [main] INFO Project target set ('dash'). Skipping non-matching project ID sw360
2021-04-23T15:24:35 [main] INFO Project target set ('dash'). Skipping non-matching project ID trace4cps
2021-04-23T15:24:35 [main] INFO Project target set ('dash'). Skipping non-matching project ID titan
```

**Comment by fredg02** [comment]
Seems like the Travis build found some syntax errors, or are these just false-positives?

**Comment by autumnfound** [comment]
Error is from a different file, crud. I'll throw a fix on top of this branch

**Review by linkfang — APPROVED** [review]
